File: sdks/go/providers/microsoft/protocol.go
package microsoft

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	out "github.com/speechswitch/client/sdks/go/generated/microsoft_output"
	"github.com/speechswitch/client/sdks/go/runtime"
	"math"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

type frame struct {
	path, requestID string
	streamID        runtime.Optional[string]
	body            []byte
	binary          bool
}

func encode(path, id, body string) runtime.WebSocketText {
	contentType := "application/json"
	if path == "ssml" {
		contentType = "application/ssml+xml"
	} else if path == "text.piece" || path == "text.end" {
		contentType = "text/plain"
	}
	return runtime.WebSocketText("Path: " + path + "\r\nX-RequestId: " + id + "\r\nX-Timestamp: " + time.Now().UTC().Format("2006-01-02T15:04:05.000Z") + "\r\nContent-Type: " + contentType + "\r\n\r\n" + body)
}

func decode(message runtime.WebSocketMessage, limit int) (frame, error) {
	f := frame{}
	var headers string
	switch v := message.(type) {
	case runtime.WebSocketText:
		if len(v) > limit {
			return f, errors.New("Microsoft message exceeds MaxMessageBytes")
		}
		head, body, found := strings.Cut(string(v), "\r\n\r\n")
		if !found {
			return f, errors.New("Microsoft text frame is missing its header separator")
		}
		headers, f.body = head, []byte(body)
	case runtime.WebSocketBinary:
		if len(v) > limit {
			return f, errors.New("Microsoft message exceeds MaxMessageBytes")
		}
		if len(v) < 2 {
			return f, errors.New("Microsoft binary frame is missing its header length")
		}
		size := int(binary.BigEndian.Uint16(v))
		if size > len(v)-2 {
			return f, errors.New("Microsoft binary frame has a truncated header")
		}
		if !utf8.Valid(v[2 : 2+size]) {
			return f, errors.New("Microsoft binary frame has invalid UTF-8 headers")
		}
		headers, f.body, f.binary = string(v[2:2+size]), v[2+size:], true
	default:
		return f, errors.New("Unsupported Microsoft WebSocket message data")
	}
	fields := map[string]string{}
	for _, line := range strings.Split(headers, "\r\n") {
		if line == "" {
			continue
		}
		name, value, found := strings.Cut(line, ":")
		name, value = strings.ToLower(strings.TrimSpace(name)), strings.TrimSpace(value)
		if !found || name == "" || strings.ContainsAny(value, "\r\n") || strings.Trim(name, "abcdefghijklmnopqrstuvwxyz0123456789-") != "" {
			return f, errors.New("Microsoft frame contains an invalid header")
		}
		if _, present := fields[name]; present {
			return f, errors.New("Microsoft frame repeats header " + name)
		}
		fields[name] = value
	}
	if fields["path"] == "" || fields["x-requestid"] == "" {
		return f, errors.New("Microsoft frame is missing Path or X-RequestId")
	}
	f.path, f.requestID = strings.ToLower(fields["path"]), fields["x-requestid"]
	if id, ok := fields["x-streamid"]; ok {
		f.streamID = runtime.Some(id)
	}
	return f, nil
}

func object(data []byte) (map[string]json.RawMessage, error) {
	if !runtime.ValidUTF8JSON(data) {
		return nil, errors.New("Microsoft returned invalid JSON")
	}
	var result map[string]json.RawMessage
	if err := json.Unmarshal(data, &result); err != nil || result == nil {
		return nil, errors.New("Microsoft returned an invalid synthesis object")
	}
	return result, nil
}

func stringValue(data json.RawMessage) (string, bool) {
	var value string
	if len(data) == 0 || data[0] != '"' {
		return "", false
	}
	err := json.Unmarshal(data, &value)
	return value, err == nil
}

func ticks(data json.RawMessage, name string) (float64, error) {
	var n float64
	if len(data) == 0 || string(data) == "null" {
		return 0, errors.New("Microsoft returned invalid " + name)
	}
	if err := json.Unmarshal(data, &n); err != nil || n < 0 || n > 9007199254740991 || math.Trunc(n) != n {
		return 0, errors.New("Microsoft returned invalid " + name)
	}
	return n, nil
}

func metadata(body map[string]json.RawMessage) ([]out.MicrosoftTimestamp, runtime.Optional[float64], error) {
	marks := []out.MicrosoftTimestamp{}
	duration := runtime.Optional[float64]{}
	var rows []json.RawMessage
	if err := json.Unmarshal(body["Metadata"], &rows); err != nil || rows == nil {
		return nil, duration, errors.New("Microsoft returned invalid synthesis Metadata")
	}
	for _, raw := range rows {
		row, err := object(raw)
		if err != nil {
			return nil, duration, err
		}
		kind, ok := stringValue(row["Type"])
		if !ok {
			return nil, duration, errors.New("Microsoft returned an invalid metadata type")
		}
		if kind != "WordBoundary" && kind != "SentenceBoundary" && kind != "Bookmark" && kind != "Viseme" && kind != "SessionEnd" {
			continue
		}
		data, err := object(row["Data"])
		if err != nil {
			return nil, duration, err
		}
		offset, err := ticks(data["Offset"], "metadata Offset")
		if err != nil {
			return nil, duration, err
		}
		if kind == "SessionEnd" {
			if duration.Present {
				return nil, duration, errors.New("Microsoft returned duplicate SessionEnd metadata")
			}
			duration = runtime.Some(offset / 10000)
			continue
		}
		mark := out.MicrosoftTimestamp{StartTimeMs: offset / 10000}
		switch kind {
		case "WordBoundary", "SentenceBoundary":
			text, err := object(data["text"])
			if err != nil {
				return nil, duration, err
			}
			value, ok := stringValue(text["Text"])
			if !ok {
				return nil, duration, errors.New("Microsoft returned invalid boundary text")
			}
			if raw, present := text["BoundaryType"]; present {
				v, ok := stringValue(raw)
				if !ok {
					return nil, duration, errors.New("Microsoft returned invalid boundary text")
				}
				mark.BoundaryType = runtime.Some(v)
			}
			length, err := ticks(data["Duration"], "metadata Duration")
			if err != nil {
				return nil, duration, err
			}
			if offset+length > 9007199254740991 {
				return nil, duration, errors.New("Microsoft metadata timing overflow")
			}
			mark.Kind = out.MicrosoftTimestampKindAsWord{}
			if kind == "SentenceBoundary" {
				mark.Kind = out.MicrosoftTimestampKindAsSentence{}
			}
			mark.Value, mark.EndTimeMs = value, runtime.Some(offset/10000+length/10000)
		case "Bookmark":
			value, ok := stringValue(data["Bookmark"])
			if !ok {
				return nil, duration, errors.New("Microsoft returned an invalid bookmark")
			}
			mark.Kind, mark.Value = out.MicrosoftTimestampKindAsSsml{}, value
		case "Viseme":
			id, err := ticks(data["VisemeId"], "viseme metadata")
			if err != nil {
				return nil, duration, err
			}
			mark.Kind, mark.Value = out.MicrosoftTimestampKindAsViseme{}, strconv.FormatFloat(id, 'f', 0, 64)
			if raw, present := data["AnimationChunk"]; present {
				v, ok := stringValue(raw)
				if !ok {
					return nil, duration, errors.New("Microsoft returned invalid viseme metadata")
				}
				mark.AnimationChunk = runtime.Some(v)
			}
			if raw, present := data["IsLastAnimation"]; present {
				var v out.MicrosoftTimestampIsLastAnimation
				switch string(raw) {
				case "true":
					v = out.MicrosoftTimestampIsLastAnimationAsTrue{}
				case "false":
					v = out.MicrosoftTimestampIsLastAnimationAsFalse{}
				default:
					return nil, duration, errors.New("Microsoft returned invalid viseme metadata")
				}
				mark.IsLastAnimation = runtime.Some(v)
			}
		}
		marks = append(marks, mark)
	}
	return marks, duration, nil
}

File: sdks/rust/src/providers/microsoft/protocol.rs
use crate::{generated::microsoft_output::*, http::TransportError, json::Raw, websocket::Message};
use std::{
    collections::BTreeMap,
    time::{SystemTime, UNIX_EPOCH},
};

#[derive(Debug, PartialEq, Eq)]
pub struct Error {
    pub message: String,
    pub status: Option<u16>,
    pub retry_after: Option<String>,
}
impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}
impl std::error::Error for Error {}
pub(super) fn failure(message: &str) -> TransportError {
    Box::new(Error {
        message: message.into(),
        status: None,
        retry_after: None,
    })
}

pub(super) fn timestamp(now: SystemTime) -> Result<String, TransportError> {
    let elapsed = now
        .duration_since(UNIX_EPOCH)
        .map_err(|_| failure("Microsoft clock precedes the Unix epoch"))?;
    let mut days = elapsed.as_secs() / 86400;
    let mut year = 1970;
    loop {
        let length = if year % 4 == 0 && (year % 100 != 0 || year % 400 == 0) {
            366
        } else {
            365
        };
        if days < length {
            break;
        }
        days -= length;
        year += 1;
        if year > 9999 {
            return Err(failure(
                "Microsoft clock exceeds the supported calendar range",
            ));
        }
    }
    let leap = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    let mut month = 1;
    for length in [
        31,
        if leap { 29 } else { 28 },
        31,
        30,
        31,
        30,
        31,
        31,
        30,
        31,
        30,
        31,
    ] {
        if days < length {
            break;
        };
        days -= length;
        month += 1;
    }
    let second = elapsed.as_secs() % 86400;
    Ok(format!(
        "{year:04}-{month:02}-{:02}T{:02}:{:02}:{:02}.{:03}Z",
        days + 1,
        second / 3600,
        second / 60 % 60,
        second % 60,
        elapsed.subsec_millis()
    ))
}
pub(super) fn encode(path: &str, id: &str, body: String) -> Result<Message, TransportError> {
    let content_type = match path {
        "ssml" => "application/ssml+xml",
        "text.piece" | "text.end" => "text/plain",
        _ => "application/json",
    };
    Ok(Message::Text(format!("Path: {path}\r\nX-RequestId: {id}\r\nX-Timestamp: {}\r\nContent-Type: {content_type}\r\n\r\n{body}",timestamp(SystemTime::now())?)))
}
pub(super) struct Frame {
    pub path: String,
    pub request_id: String,
    pub stream_id: Option<String>,
    pub body: Message,
}
pub(super) fn decode(message: Message, limit: usize) -> Result<Frame, TransportError> {
    let (headers, body) = match message {
        Message::Text(mut data) => {
            if data.len() > limit {
                return Err(failure("Microsoft message exceeds max_message_bytes"));
            }
            let end = data
                .find("\r\n\r\n")
                .ok_or_else(|| failure("Microsoft text frame is missing its header separator"))?;
            let body = data.split_off(end + 4);
            data.truncate(end);
            (data, Message::Text(body))
        }
        Message::Binary(mut data) => {
            if data.len() > limit {
                return Err(failure("Microsoft message exceeds max_message_bytes"));
            }
            if data.len() < 2 {
                return Err(failure(
                    "Microsoft binary frame is missing its header length",
                ));
            }
            let length = u16::from_be_bytes([data[0], data[1]]) as usize;
            if length > data.len() - 2 {
                return Err(failure("Microsoft binary frame has a truncated header"));
            }
            let body = data.split_off(length + 2);
            let headers = std::str::from_utf8(&data[2..])
                .map_err(|_| failure("Microsoft binary frame has invalid UTF-8 headers"))?
                .to_owned();
            (headers, Message::Binary(body))
        }
    };
    let mut fields = BTreeMap::new();
    for line in headers.split("\r\n").filter(|v| !v.is_empty()) {
        let (name, value) = line
            .split_once(':')
            .ok_or_else(|| failure("Microsoft frame contains an invalid header"))?;
        let (name, value) = (name.trim().to_ascii_lowercase(), value.trim().to_owned());
        if name.is_empty()
            || !name
                .bytes()
                .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
            || value.contains(['\r', '\n'])
        {
            return Err(failure("Microsoft frame contains an invalid header"));
        }
        if fields.insert(name.clone(), value).is_some() {
            return Err(failure(&format!("Microsoft frame repeats header {name}")));
        }
    }
    let path = fields
        .remove("path")
        .filter(|v| !v.is_empty())
        .ok_or_else(|| failure("Microsoft frame is missing Path or X-RequestId"))?;
    let request_id = fields
        .remove("x-requestid")
        .filter(|v| !v.is_empty())
        .ok_or_else(|| failure("Microsoft frame is missing Path or X-RequestId"))?;
    Ok(Frame {
        path: path.to_ascii_lowercase(),
        request_id,
        stream_id: fields.remove("x-streamid"),
        body,
    })
}
pub(super) fn object(text: &str) -> Result<BTreeMap<String, Raw<'_>>, TransportError> {
    Raw::parse(text)
        .map_err(|_| failure("Microsoft returned invalid JSON"))?
        .object()
        .map_err(|_| failure("Microsoft returned an invalid synthesis object"))
}
fn ticks(value: Option<Raw<'_>>, name: &str) -> Result<f64, TransportError> {
    value
        .and_then(|v| v.number().ok())
        .filter(|n| n.is_finite() && *n >= 0.0 && *n <= 9007199254740991.0 && n.fract() == 0.0)
        .ok_or_else(|| failure(&format!("Microsoft returned invalid {name}")))
}
pub(super) fn metadata(
    body: BTreeMap<String, Raw<'_>>,
) -> Result<(Vec<MicrosoftTimestamp>, Option<f64>), TransportError> {
    let rows = body
        .get("Metadata")
        .and_then(|v| v.array().ok())
        .ok_or_else(|| failure("Microsoft returned invalid synthesis Metadata"))?;
    let mut marks = vec![];
    let mut duration = None;
    for row in rows {
        let row = row
            .object()
            .map_err(|_| failure("Microsoft returned an invalid synthesis object"))?;
        let kind = row
            .get("Type")
            .and_then(|v| v.string().ok())
            .ok_or_else(|| failure("Microsoft returned an invalid metadata type"))?;
        if ![
            "WordBoundary",
            "SentenceBoundary",
            "Bookmark",
            "Viseme",
            "SessionEnd",
        ]
        .contains(&kind.as_str())
        {
            continue;
        }
        let data = row
            .get("Data")
            .and_then(|v| v.object().ok())
            .ok_or_else(|| failure("Microsoft returned an invalid synthesis object"))?;
        let offset = ticks(data.get("Offset").copied(), "metadata Offset")?;
        if kind == "SessionEnd" {
            if duration.is_some() {
                return Err(failure("Microsoft returned duplicate SessionEnd metadata"));
            };
            duration = Some(offset / 10000.0);
            continue;
        }
        let mut mark = MicrosoftTimestamp {
            kind: MicrosoftTimestampKind::Ssml(MicrosoftTimestampKindSsml),
            value: String::new(),
            start_time_ms: offset / 10000.0,
            end_time_ms: None,
            source: None,
            boundary_type: None,
            animation_chunk: None,
            is_last_animation: None,
        };
        match kind.as_str() {
            "WordBoundary" | "SentenceBoundary" => {
                let text = data
                    .get("text")
                    .and_then(|v| v.object().ok())
                    .ok_or_else(|| failure("Microsoft returned an invalid synthesis object"))?;
                mark.value = text
                    .get("Text")
                    .and_then(|v| v.string().ok())
                    .ok_or_else(|| failure("Microsoft returned invalid boundary text"))?;
                if let Some(value) = text.get("BoundaryType") {
                    mark.boundary_type = Some(
                        value
                            .string()
                            .map_err(|_| failure("Microsoft returned invalid boundary text"))?,
                    );
                }
                let length = ticks(data.get("Duration").copied(), "metadata Duration")?;
                if offset + length > 9007199254740991.0 {
                    return Err(failure("Microsoft metadata timing overflow"));
                }
                mark.kind = if kind == "WordBoundary" {
                    MicrosoftTimestampKind::Word(MicrosoftTimestampKindWord)
                } else {
                    MicrosoftTimestampKind::Sentence(MicrosoftTimestampKindSentence)
                };
                mark.end_time_ms = Some(offset / 10000.0 + length / 10000.0);
            }
            "Bookmark" => {
                mark.value = data
                    .get("Bookmark")
                    .and_then(|v| v.string().ok())
                    .ok_or_else(|| failure("Microsoft returned an invalid bookmark"))?;
            }
            "Viseme" => {
                let id = ticks(data.get("VisemeId").copied(), "viseme metadata")?;
                mark.kind = MicrosoftTimestampKind::Viseme(MicrosoftTimestampKindViseme);
                mark.value = id.to_string();
                if let Some(value) = data.get("AnimationChunk") {
                    mark.animation_chunk = Some(
                        value
                            .string()
                            .map_err(|_| failure("Microsoft returned invalid viseme metadata"))?,
                    );
                }
                if let Some(value) = data.get("IsLastAnimation") {
                    mark.is_last_animation = Some(
                        if value
                            .boolean()
                            .map_err(|_| failure("Microsoft returned invalid viseme metadata"))?
                        {
                            MicrosoftTimestampIsLastAnimation::True(
                                MicrosoftTimestampIsLastAnimationTrue,
                            )
                        } else {
                            MicrosoftTimestampIsLastAnimation::False(
                                MicrosoftTimestampIsLastAnimationFalse,
                            )
                        },
                    );
                }
            }
            _ => unreachable!(),
        }
        marks.push(mark);
    }
    Ok((marks, duration))
}
